### PR TITLE
fix(issue): execute-task post-task commit silently swallows deterministic pre-commit-hook failures, strands task work uncommitted

### LINKED ISSUE #1491: execute-task post-task commit silently swallows deterministic pre-commit-hook failures, strands task work uncommitted
## Problem

For `execute-task` units, GSD's post-task auto-commit silently swallows deterministic commit failures (e.g. a pre-commit hook rejecting the staged content) and advances to the next task with the previous task's results left uncommitted in the working tree. The user expects GSD to feed the hook's stderr back to the model so it can remedy the issue and finalize the commit; instead the work is stranded and only surfaces later (often as a dirty-worktree quarantine on `stopAuto` — see companion issue).

## Root Cause

Three compounding defects in the post-unit commit path:

### Defect 1 — `execute-task` commits are classified as soft-failure

`auto-post-unit.js:531-533`:
```js
export function shouldDeferCloseoutGitAction(unitType) {
    return unitType === "execute-task";
}
```
This defers the `execute-task` closeout commit out of the blocking pre-verification path (the blocking `runCloseoutGitAction` call at `auto-post-unit.js:1091` is skipped) and into `postUnitPostVerification`, where it runs with the soft flag (`auto-post-unit.js:1910-1915`):
```js
if (shouldDeferCloseoutGitAction(s.currentUnit.type)) {
    const gitActionResult = await runCloseoutGitAction(pctx, s.currentUnit, { softFailure: true });
    if (gitActionResult === "dispatched") { return "stopped"; }
}
```

### Defect 2 — "soft failure" means "give up and continue", not "remedy"

`auto-post-unit.js:921` and `:930`:
```js
const maxAttempts = opts?.softFailure ? 3 : 1;
...
for (let attempt = 1; gitResult.status === "failed" && attempt < maxAttempts; attempt++) {
    await new Promise((resolve) => setTimeout(resolve, 250 * attempt));
    gitResult = runTurnGitAction({ ... });   // same staged content, same failure
}
```
`auto-post-unit.js:1001-1003`:
```js
if (opts?.softFailure) {
    return "continue";          // ← next task dispatches with previous task uncommitted
}
```
The three retries only help for *transient* failures (index lock, fs race). For **deterministic** pre-commit-hook failures (lint/format/secret/test), every retry re-stages identical content and fails identically. After the loop, `return "continue"` lets the dispatcher pick the next task — the reported symptom.

### Defect 3 — the inner retry in `autoCommit` only handles one specific hook behavior

`git-service.js:794-805`:
```js
try {
    nativeCommit(this.basePath, message, { allowEmpty: false });
} catch (err) {
    // Some pre-commit hooks intentionally rewrite files and fail the first
    // commit to force a re-stage + retry.
    if (!nativeHasChanges(this.basePath)) throw err;
    const retriedScoped = taskContext ? this.scopedStageTaskFiles(taskContext, extraExclusions) : false;
    if (!retriedScoped) this.smartStage(extraExclusions);
    if (!nativeHasStagedChanges(this.basePath)) throw err;
    nativeCommit(this.basePath, message, { allowEmpty: false });   // throws again → propagates
}
```
The comment betrays the assumption: the retry exists solely for hooks that *rewrite files on first attempt*. Hooks that fail because the **content is wrong** (the model needs to fix it) re-throw on the second `nativeCommit`, propagating up to Defect 2's soft-continue.

### Contrast with the correct pattern

`runCloseoutGitAction` *does* have the right shape for non-soft failures (`auto-post-unit.js:1001-1007`): it sets `s.lastGitActionFailure`, persists a failure log via `persistGitActionFailure`, runs a `UokGateRunner` gate, and calls `pauseAuto(ctx, pi)` + `return "dispatched"` to halt the loop. **None of that remediation logic fires for `execute-task` because of Defect 1.**

## Expected Behavior

When a post-task commit fails on an `execute-task` and the failure is deterministic (pre-commit hook output present, not a transient lock error), GSD should:

1. **Feed the hook's stderr back to the model via a remediation re-dispatch** of the same task, with the commit error injected as failure context (the same mechanism already used for failed artifact verification — `s.pendingVerificationRetry`). This is the "remedy the issues and finalize the commit" behavior users expect.
2. **Cap remediation attempts** (e.g. 2) before falling back to `pauseAuto` — don't loop forever on a hook the model genuinely can't satisfy.
3. **Stop classifying `execute-task` commit failures as universally soft.** `shouldDeferCloseoutGitAction` (`auto-post-unit.js:531`) is the wrong gate; the soft-vs-blocking decision should be made from the *failure class* (transient vs. deterministic content failure), not the *unit type*.
4. **Distinguish failure classes in `runTurnGitAction`'s result** (`git-service.js:1182`) — add a `failureClass: "transient" | "hook-content" | "unknown"` so the closeout path knows whether to retry, re-dispatch the model, or pause.

## Environment

- GSD version: 1.11.0
- Unit type: `execute-task` (reproduced across multiple milestones)

## Reproduction Context

M056/S01 "Discord OAuth and Platform Registry" — a 7-task slice produced only **4 commits** on the `milestone/M056` branch. Tasks T05 (Discord OAuth tests), T06 (Discord UI), and T07 (signup attribution reporting) each completed — full summaries written, all verification green — but their post-task commits were defeated by a pre-commit gitleaks hook false-positive on the synthetic Discord snowflake `123456789012345678` used in 5 test fixtures. The `discord-client-id` built-in gitleaks rule flagged the test data as a secret; the hook hard-failed; GSD's soft-failure path swallowed it; the next task dispatched. The uncommitted work for all three tasks was only preserved because the dirty-worktree quarantine fired on `stopAuto` (see companion issue).

`milestone/M056` reflog at recovery time:
```
55300137 (recovery commit, after manual fix) — T05+T06+T07 + gitleaks:allow directives
6a56db94 chore: Added Discord as the first platform in config/platforms.php — T04
325e1730 feat: Widened OAuthController's provider allowlist — T03
c1e40f17 chore: Added discord key to config/services.php — T02
feb1e340 feat: Added socialiteproviders/discord dependency — T01
cef58b41 branch: Created from main
```

## Forensic Evidence

- Forensic report flagged 60+ `[WARNING] stuck-loop` anomalies on `execute-task/M###/S##/T##` units — consistent with: task completes, commit fails softly, working tree stays dirty, downstream artifact verification re-dispatches the unit because the expected commit artifact isn't there.
- Quarantine metadata for the recovered M056 worktree listed **30 staged/modified files** (the uncommitted T05-T07 work) despite all three tasks being marked complete in the DB.
- After adding `// gitleaks:allow` directives to the 5 flagged test-fixture lines and committing manually, all 152 affected tests pass (424 assertions) — confirming the task work itself was correct; only the commit pipeline failed.
- The companion issue documents the secondary data-loss-risk path this creates at `stopAuto`.

## Recovery (immediate unstuck for affected users)

In the affected repo:
```bash
git status                       # see the uncommitted previous-task changes
git add -A && git commit -m "…"  # commit manually; run the pre-commit hook manually to see the real error
# fix whatever the hook complains about, then re-commit
```
Then resume auto-mode. To prevent recurrence until the fix lands: temporarily disable the offending pre-commit hook, or set `git.turn_action: "status-only"` so commits happen manually.

---
*Auto-generated by `/gsd forensics`*

### PR DESCRIPTION
**Pull request by jeremymcs**
## Summary
- Routed deterministic execute-task commit hook failures into bounded remediation retries and verified diff and TypeScript syntax checks; full tests were blocked by missing dependencies and restricted network.

## Verification
- Completed in the repository worktree before push.

## Related Issue
- Closes #1491
- [#1491 execute-task post-task commit silently swallows deterministic pre-commit-hook failures, strands task work uncommitted](https://github.com/open-gsd/gsd-pi/issues/1491)

## User
- @vburghelea

## Repo
- `open-gsd/gsd-pi`

## Branch
- `issue/1491-execute-task-post-task-commit-silently-s-1784416060`

---

> [!NOTE]
> **Medium Risk**
> Changes auto-mode post-verification git closeout and execute-task re-dispatch after DB completion; misclassification could pause too often or retry incorrectly, but behavior is bounded and covered by new integration tests.
> 
> **Overview**
> **Post-task `git commit` failures are classified and handled by cause** instead of treating most closeout commit errors as ignorable soft warnings.
> 
> Turn git actions now attach a `failureClass` (`transient`, `hook-content`, or `unknown`). Lock/contention errors keep the short retry loop; only **transient** failures may warn-and-continue under deferred closeout. Pre-commit and other hook rejections on verified `execute-task` units are **`hook-content`**: auto mode injects the full hook output into `pendingVerificationRetry`, re-dispatches the same task (up to **2** attempts), and pauses with `.gsd/git-action-failures.log` if remediation is exhausted.
> 
> **Dispatch and orchestration** no longer skip an already-`complete` task when the retry signature is `git-commit:*`, so remediation can run after verification already closed the task in the DB. **Parent workspaces** with a partial multi-repo commit (some repos committed, another hook failed) **pause** without task redo to avoid duplicating committed work.
> 
> User docs and GitBook describe the new remediation path and troubleshooting steps. Tests cover classification edge cases, remediation cap, partial commits, and orchestrator bypass.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 047d1c9961699a23f6a8d9a823ebc0a40b4ff810. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
<a href="https://app.blacksmith.sh/open-gsd/codesmith/gsd-pi/pr/1497"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is enabled.</sup>

### CODE REVIEW

**Comment by jeremymcs** [comment]
✅ **Verification — all 1 task is addressed**

Re-checked PR #1497 against [#1491 execute-task post-task commit silently swallows deterministic pre-commit-hook failures, strands task work uncommitted](https://github.com/open-gsd/gsd-pi/issues/1491).

## Subtasks
- [x] **execute-task post-task commit silently swallows deterministic pre-commit-hook failures, strands task work uncommitted**

_Re-run from Patchdeck to refresh._

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/git-service.ts [hunk 1]
```diff
@@ -1328,10 +1330,41 @@ export function handleTurnGitActionError(action: TurnGitActionMode, err: unknown
...
+  const stderr = errorWithStreams.stderr?.trim() ?? "";
+  const message = errorWithStreams.message ?? getErrorMessage(err);
+  const combined = `${stderr}\n${message}`;
+  if (TRANSIENT_GIT_FAILURE_PATTERNS.some((pattern) => pattern.test(combined))) {
+    return "transient";
+  }
+  if (action === "commit" && stderr && errorWithStreams.status === 1) {
+    return "hook-content";
+  }
+  return "unknown";
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1382]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `2bf8a86`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `2bf8a86`.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `564a2be`.

File: src/resources/extensions/gsd/git-service.ts [hunk 2]
```diff
@@ -1328,10 +1330,41 @@ export function handleTurnGitActionError(action: TurnGitActionMode, err: unknown
...
+  const errorWithStreams = err as { stderr?: string; message?: string; status?: number };
+  const stderr = errorWithStreams.stderr?.trim() ?? "";
+  const message = errorWithStreams.message ?? getErrorMessage(err);
+  const combined = `${stderr}\n${message}`;
+  if (action === "commit" && stderr && errorWithStreams.status === 1) {
+    return "hook-content";
+  }
+  if (TRANSIENT_GIT_FAILURE_PATTERNS.some((pattern) => pattern.test(combined))) {
+    return "transient";
+  }
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1381]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `2bf8a86`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `2bf8a86`.

File: src/resources/extensions/gsd/auto-post-unit.ts [hunk 3]
```diff
@@ -1323,21 +1333,63 @@ async function runCloseoutGitAction(
...
+            );
+            debugLog("postUnit", {
+              phase: "git-action-remediation-retry",
+              action: turnAction,
+              unitType: unit.type,
+              unitId: unit.id,
+              attempt,
+              failureLogPath,
+            });
+            return "retry";
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-post-unit.ts:1376]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `2bf8a86`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/git-service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1382]: Addressed in commit `2bf8a86`.

Updated commit failure classification so exit-1 hook stderr with lock-like user text remains `hook-content`, avoiding the transient warn-and-continue path for hook rejections.

**Comment by jeremymcs** [comment]
Addressed in commit `2bf8a86`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1497#pullrequestreview-4729494845):
> 

Addressed the hook misclassification by preserving hook-content routing for non-git-owned exit-1 commit stderr and verified it with focused regression tests.

File: src/resources/extensions/gsd/git-service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1381]: Addressed in commit `2bf8a86`.

Added a git-owned transient lock detector before the exit-1 hook fallback, so real git lock output such as `.git/index.lock` is classified as `transient` even when status is `1`.

**Comment by jeremymcs** [comment]
Addressed in commit `2bf8a86`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1497#pullrequestreview-4729497441):
> 

Addressed transient lock misclassification with a new exit-1 lock regression that verifies git-owned lock errors route to transient retries instead of hook remediation.

File: src/resources/extensions/gsd/auto-post-unit.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-post-unit.ts:1376]: Addressed in commit `2bf8a86`.

Verified the current partial-commit guard skips execute-task remediation when `commitMessages` show some repositories already committed, and reran the partial multi-repo closeout regression successfully.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/auto-post-unit.ts [hunk 4]
```diff
@@ -1323,21 +1333,70 @@ async function runCloseoutGitAction(
...
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-post-unit.ts:1376]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `564a2be`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/git-service.ts [hunk 5]
```diff
@@ -1328,10 +1330,54 @@ export function handleTurnGitActionError(action: TurnGitActionMode, err: unknown
...
+  return "unknown";
+}
+
+function mergeTurnGitFailureClasses(classes: readonly TurnGitFailureClass[]): TurnGitFailureClass {
+  if (classes.includes("unknown")) return "unknown";
+  if (classes.includes("hook-content")) return "hook-content";
+  if (classes.includes("transient")) return "transient";
+  return "unknown";
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1389]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Resolved** — addressed in commit `564a2be`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/auto-post-unit.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-post-unit.ts:1376]: Addressed in commit `564a2be`.

**Comment by jeremymcs** [comment]
Addressed in commit `564a2be`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1497#pullrequestreview-4729683327):
>

File: src/resources/extensions/gsd/git-service.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1389]: Addressed in commit `564a2be`.

The branch already had `mergeTurnGitFailureClasses` prioritizing `hook-content` over `unknown`; I added a focused parent-workspace regression test covering mixed hook-content and unknown failures so bounded task remediation remains reachable.

**Comment by jeremymcs** [comment]
Addressed in commit `564a2be`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1497#pullrequestreview-4729689603):
> 

Addressed the Bugbot review issue by preserving the existing merge precedence fix and adding verification coverage for the multi-repo failure case; the focused git action test suite passes.

**Comment by jeremymcs** [comment]
## CI needs attention

Changes were pushed in commit `564a2be`, but CI/CD checks are still failing:

- **node22-smoke**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670226129/job/88148016994))
- **windows-portability**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670226129/job/88148016780))
- **ci-gate**: Check run failure ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670226129/job/88148016698))
- **build**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670226129/job/88147765650))

Review the failing checks before merging.

File: src/resources/extensions/gsd/git-service.ts [hunk 6]
```diff
@@ -1328,10 +1330,54 @@ export function handleTurnGitActionError(action: TurnGitActionMode, err: unknown
...
+function classifyTurnGitActionFailure(action: TurnGitActionMode, err: unknown): TurnGitFailureClass {
+  const errorWithStreams = err as { stderr?: string; message?: string; status?: number };
+  const stderr = errorWithStreams.stderr?.trim() ?? "";
+  const message = errorWithStreams.message ?? getErrorMessage(err);
+  const combined = `${stderr}\n${message}`;
+  if (action === "commit" && stderr !== "" && errorWithStreams.status === 1) {
+    if (GIT_OWNED_TRANSIENT_GIT_FAILURE_PATTERNS.some((pattern) => pattern.test(stderr))) {
+      return "transient";
+    }
+    return "hook-content";
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1377]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `2b2c365`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/git-service.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/git-service.ts:1377]: Addressed in commit `2b2c365`.

No additional code changes were needed because current HEAD already classifies commit hook stderr without a numeric status as `hook-content`; verified with the focused gitops test file and a direct missing-status classifier assertion.

**Comment by jeremymcs** [comment]
## CI needs attention

Changes were pushed in commit `2b2c365`, but CI/CD checks are still failing:

- **windows-portability**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670323858/job/88148436011))
- **node22-smoke**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670323858/job/88148435971))
- **ci-gate**: Check run failure ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670323858/job/88148435905))
- **build**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/29670323858/job/88148056161))

Review the failing checks before merging.
